app/js/lib/foxbox.js

```diff
@@ -218,78 +246,57 @@ export default class Foxbox extends Service {
-          // Detect change in service states.
-          const storedServices = res[0];
-          const fetchedServices = res[1];
-
-          Promise.all(fetchedServices.map(service => this.getServiceState(service.id)))
-            .then(states => {
-              fetchedServices.forEach((fetchedService, id) => {
-                // Populate the service objects with states.
-                fetchedService.state = states[id];
-
-                const storedService = storedServices.find(s => s.id === fetchedService.id);
-
-                if (!storedService) {
-                  this._dispatchEvent('service-state-change', fetchedService);
-
-                  // Populate the db with the latest service.
-                  db.setService(fetchedService);
+    let fetchedServicesPromise = fetchJSON(`${this.origin}/services/list`).then(
+      (services) => {
+        // TODO: We should ask for state only for services that actually support
+        // it.
+        return Promise.all(
+          services.map((service) => {
+            // Use empty state if service failed to return actual state.
+            return this.getServiceState(service.id).catch(() => ({})).then(
+              (state) => service.state = state
+            );
+          })
+        ).then(() => services);
+      }
+    );
+
+    return Promise.all([
+      this.getServices(),
+      fetchedServicesPromise
+    ]).then(([storedServices, fetchedServices]) => {
+      let hasNewServices = fetchedServices.reduce(
+        (hasNewServices, fetchedService) => {
+          const storedService = storedServices.find(
+            s => s.id === fetchedService.id
+          );
+
+          const isExistingService = !!storedService;
+
+          if (isExistingService &&
+              isSimilar(fetchedService.state, storedService.state)) {
+            return hasNewServices;
+          }
 
-                  return;
-                }
+          fetchedService = isExistingService ?
+            Object.assign(storedService, fetchedService) : fetchedService;
 
-                if (!isSimilar(fetchedService.state, storedService.state)) {
-                  fetchedService = Object.assign(storedService, fetchedService);
+          this._dispatchEvent('service-state-change', fetchedService);
 
-                  this._dispatchEvent('service-state-change', fetchedService);
+          // Populate the db with the latest service.
+          db.setService(fetchedService);
 
-                  // Populate the db with the latest service.
-                  db.setService(fetchedService);
-                }
-              });
+          return hasNewServices || !isExistingService;
+        },
+        false /* hasNewServices */
+      );
 
-              if (hasChanged) {
-                // The state of the services changes.
-                this._dispatchEvent('service-change', fetchedServices);
-              }
+      if (hasNewServices || fetchedServices.length !== storedServices.length) {
```

Comment:
I think so, let me think aloud:
* If new service connected we'll have ```hasNewServices === true``` since we iterate through all _latest_ services and if some are new - we'll have this flag set, independently from any disconnected services;
* If old service disconnected, we'll have different length;
* If new service connected and old is disconnected we'll have ```hasNewServices === true``` anyway.

Am I missing something?